src/utils/posixifyPathBuffer.js

```diff
@@ -0,0 +1,8 @@
+export function posixifyPathBuffer(buffer) {
+  if (buffer instanceof Buffer) {
+    let idx
+    while (~(idx = buffer.indexOf(92))) buffer[idx] = 47
+    return buffer
+  }
+  return Buffer.from(buffer.startsWith('./') ? buffer.substr(2) : buffer)
+}
```

Comment:
I think that the place where it's added it not optimal. I would just add this to [FileSystem proxy](https://github.com/isomorphic-git/isomorphic-git/blob/main/src/models/FileSystem.js). That is a facade for the `fs` client. You can there return Buffer like in [read method](https://github.com/isomorphic-git/isomorphic-git/blob/main/src/models/FileSystem.js#L66).

You should do all manipulation in FileSystem object.

Also, I have a question: why you don't use `Buffer.from(buffer.replace(/\\/g, '/'))` on a string the same as what you do with buffer?

Reply:
> why you don't use Buffer.from(buffer.replace(/\\/g, '/')) on a string the same as what you do with buffer?

When the type is a buffer, I manipulated it directly as a performance optimization.

Reply:
> I think that the place where it's added it not optimal. I would just add this to FileSystem proxy. That is a facade for the fs client. You can there return Buffer like in read method.

In this case, I don't think that's the right choice. The reason is, it's not wrong that the fs.readlink returns a path with the system-dependent separator. But it is wrong for git.add. So we are modifying it at the exact boundary of where it should be modified. So I don't agree this should be done by the FileSystem object.

Reply:
(Usually, I am the one making the argument for using FileSystem object. But it's just not appropriate in this case).

What is wrong is that LightningFS should be returning a Buffer, not a string. And it should not be adding `./` to the start of that string. But those are bugs in LightningFS and we need to patch them until a new version of LightningFS is released. Then we can come back and remove the workaround.

Reply:
I agree with @mojavelinux that the `FileSystem` proxy should not contain this behavior. `FileSystem` should not have any knowledge about git IMHO. Adding the transformation the the proxy will impose git-specific behavior onto it, making it harder to re-use in other use-cases.

I've got a different concern with this particular piece of code though and that is that the behavior (removing `./`  __or__ removing backslashes) is controlled by the type of `buffer` that is passed into the function. I get that if it's a string, we're dealing with LightningFS and want to treat it differently. It just seems to me that the fact LightningFS uses a string instead of a buffer is coincidental and not clear to the reader. IMHO the code would be more readable if we'd first check the type, convert it to buffer if it's a string and then apply the replacements, regardless of which specific fs implementation is used.

It just doesn't seem right to me to detect the fs implementation based on this bug.

Maybe a solution would be adding the type conversion into the proxy, so that buffers are always returned by `fs.readlink()`. The logic for dealing with backslashes and prefixes would still be separate, but operate on buffers. Also, the replacement logic could be written in such a way that it can deal with all known issues and not assume the presence of certain bugs in the fs implementaion.

Reply:
> Maybe a solution would be adding the type conversion into the proxy, so that buffers are always returned by fs.readlink(). The logic for dealing with backslashes and prefixes would still be separate, but operate on buffers. Also, the replacement logic could be written in such a way that it can deal with all known issues and not assume the presence of certain bugs in the fs implementation.

This is exactly what I've written in the previous message but with different words.

> I think that returning Buffer from readlink should be done FileSystem object

And before patching I would check what exactly is the result of fs.readlink on WIndows and check if lightingFS return correct results, if the path is posix like then I would patch lighintgFS and if not then I will check how canonical git does this
on Windows, and do the same.

Also, Note that FileSystem object is only for git stuff, it's a wrapper for fs client. FileSystem is not suppose to be used bot anyone except git.

Reply:
But if the patch is only needed because of a bug in lightingFS I would fix lightingFS instead of introducing this temporary patch that will be removed anyway.

Reply:
> This is exactly what I've written in the previous message but with different words.

Sorry @jcubic, didn't see your message. Am a bit puzzled why the order of messages is messed up. Your message shows as "7 minutes ago" for me and my message as "1 hour ago". Full agree on the points you raised though.

Reply:
It seems that we just found a bug in GitHub. I've added one comment you added later and I was not able to comment to your comment and by mistake added my reply as a normal comment not in review and my comment that you've missed is in two places with same time.

Reply:
It seems that we just found a bug in GitHub. I've added one comment you added later and I was not able to comment to your comment and by mistake added my reply as a normal comment not in review and my comment that you've missed is in two places with same time.

Reply:
It seems that we just found a bug in GitHub. My comment that you've missed is in two places at the same time.

Reply:
> it return 3 completely different not compatible results:

I'm not trying to make a generic function. I'm trying to make a function that gets the job done it needs to do. And it does that, as the tests prove.

Reply:
I think we are bikeshedding the heck out of this problem. The code is simple and the tests pass. What's the motivation for blocking this? We need to fix the problem in isomorphic-git.

Okay, so LightningJS has a bug. If that's so much of a problem, let's just remove the workaround for it and say you can't use LightningJS until the bug in it is fixed. Frankly, I don't care about. I was just trying to be a good citizen of isomorphic-git. What I care about is Node.js.

Reply:
> Why not add path fix to lightingFS, it will make much more sense.

Because that is outside the scope of what I'm willing to commit time on right now. I've already spend 2 whole days on this patch...just to switch backslashes to forward slashes.

Reply:
I'm sorry this turned out so frustrating your you @mojavelinux. FWIW, I just saw that this PR is open for a couple of days and wanted to give some input. 

I'd also like to see this fixed and frankly, I don't think adding workarounds for 3rd party packages is in the scope of this project.

What irks me is that the implementation provided only strips out backslashes when a buffer is passed. This doesn't seems architecturally sound to me, but I also concede this might be the simplest way to fix this at the moment. 

My worry is that this function will become another difficult to understand part in an already somewhat confusing codebase. I say at least add a comment why this works.

Reply:
> What irks me is that the implementation provided only strips out backslashes when a buffer is passed.

It's only ever supposed to be a Buffer. I could add a comment as to why there is a fallback with a note that it should be removed when LightningFS is fixed...or, like I suggested, we can just get rid of it and not worry about LightningFS.

Reply:
> It's only ever supposed to be a Buffer.

Right. And it's also not supposed to start with `./`, correct? I'd be okay with handling backslashes as the default case and handling anything else as a fallback with a note describing the case (the type being string, the prefix and that it can't contain backslashes by nature).

These are just my two cents though, @jcubic is calling the shots.

Reply:
I don't know how popular LightningFS is, but removing it seems a little radical to me.

Reply:
When it is not a Buffer, we know 100% that it's coming from LightningFS since it's the only fs not conforming to the spec by honoring the encoding option. (See https://github.com/isomorphic-git/isomorphic-git/blob/04fa3d9922783341ce2e4c5a45c4a9b3db69dd17/src/models/FileSystem.js#L201). That's where we could handle the translation to a Buffer (and perhaps even deal with removing the invalid leading `./`).

I'm willing to make a change, but I need some clear direction.

Reply:
> I don't know how popular LightningFS is, but removing support for it seems a little radical to me.

I'm not saying removing LightningFS outright. I'm saying removing the workaround that would enable it to pass the assertion in the test. (As it turns out, the leading `./` is benign if the value is passed to path.join because it will get cleaned up then...but it breaks our assertion). I could even update the assertion to allow it to be there...and thus just remove this whole workaround.

Reply:
> That's where we could handle the translation to a Buffer (and perhaps even deal with removing the invalid leading ./).

I think this is a good place to deal with the LightningFS specific things.

Removing the backslashes could be done in your helper still. I think this is also what @jcubic suggested before.

Reply:
> I'm saying removing the workaround that would enable it to pass the assertion in the test.

I'd like to ultimately see the workaround removed from the project. As it stands right now, it is probably more beneficial to have it, so that we don't break any existing integrations. We should probably keep an eye out on LightningFS and see if this gets fixed. 

To reiterate, I think the proxy is a good place for this workaround as it separates the fix you wanted to contribute originally from LightningFS weirdness.

Reply:
I updated the code. See what you think now.

I am still coercing the return value of readlink to a Buffer, but doing that part in FileSystem. That way, it adapts to the incorrect return value by LightningFS is a way that is very benign and forgiving. We can remove this eventually, but for now it allows symlinks to work when using LightningFS (which would otherwise crash on byteLength property not found due to the incorrect type).

Reply:
Can I use `.then` in this case?

```js
return this._readlink(filename, opts).then(link => Buffer.isBuffer(link) ? link : Buffer.from(link))
```

I ask since `then` is not used anywhere else in that file...so I'm fine with just keeping it as it is.

Reply:
I like it 👍

Thanks for not bailing on this rather lengthy discussion 😅

I believe this is small change improves comprehensibility a lot and while it took us some time to arrive at a consensus, I think it was worthwhile. 

Code is read many more times than it is being read and it's important we get this right and improve maintainability moving forward. 

Let's see what @jcubic thinks 🤔

Reply:
> Can I use .then in this case?

Seems to me that all of these internal methods are promisified, so yes, you should be able to use `.then` here.

Reply:
Thanks for working to get to that consensus.

>  so yes, you should be able to use .then here.

I was asking about code style. Is it acceptable or do we prefer the use of await with a local const?

Reply:
This looks great. BTW: You've said in the previous comment that it was ok because tests were passing, but it was the problem with tests that they didn't check all the cases that I've shown in the comment. Now that the code is always executed I think it's fine.

But what about starting `./` is that not the problem if it's included? In the first code, you remove that from the string. Does it need to be removed in buffer?

Reply:
After looking into more, I was wrong that the extra leading `./` breaks the usage of isomorphic-git, even though it is clearly different than what is returned in other environments. Therefore, I updated the the assertion to allow for it.